File: dp/triangle.py
def solution(triangle):
	answer = 0
	h = len(triangle)
	triangle.reverse()

	B = [[0]*(h-1) for i in range(h-1)]

	for i in range(h):
		if i == 0 :
			for j in range(h-i-1):
				left = triangle[i][j] + triangle[i+1][j]
				right = triangle[i][j+1] + triangle[i+1][j]
				B[i][j] = max(left,right)
		else :
			for j in range(h-i-1):
				left =	B[i-1][j] + triangle[i+1][j]
				right = B[i-1][j+1] + triangle[i+1][j]
				B[i][j] = max(left,right)
	
	return B[h-2][0]
	


b = [[7], [3, 8], [8, 1, 0], [2, 7, 4, 4], [4, 5, 2, 6, 5]]

a = solution(b)
print (a)

File: dp/triangle.java
import java.util.*;

class Solution {
    public int solution(int[][] triangle) {
        int answer = 0;
        int n = triangle.length;
        int[][] dp = new int[n][n];
        dp[0][0] = triangle[0][0];
        for(int i = 1; i < n; i++){
            for(int j = 0; j <= i; j++){
                if(j == 0){
                    dp[i][j] = triangle[i][j] + dp[i-1][j];
                }
                else if(j == i){
                    dp[i][j] = triangle[i][j] + dp[i-1][j-1];
                } else {
                    dp[i][j] = Math.max(dp[i-1][j],dp[i-1][j-1]) + triangle[i][j];
                }
            }
        }
        for(int e : dp[n-1]){
            answer = Math.max(answer,e);
        }
        
        return answer;
    }
}
